plugins/intel-gsc/fu-igsc-aux-device.c

```diff
@@ -0,0 +1,205 @@
+		return NULL;
+
+	/* search the device list for a match */
+	if (!fu_igsc_aux_firmware_match_device(firmware,
+					       self->major_version,
+					       self->major_vcn,
+					       fu_igsc_device_get_ssvid(igsc_parent),
+					       fu_igsc_device_get_ssdid(igsc_parent),
+					       error))
+		return NULL;
+
+	/* verify is compatible */
+	if (fu_igsc_aux_firmware_get_major_version(firmware) != self->major_version) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "image is not for this product, got 0x%x, expected 0x%x",
+			    fu_igsc_aux_firmware_get_major_version(firmware),
+			    self->major_version);
+		return NULL;
+	}
+	if (fu_igsc_aux_firmware_get_major_vcn(firmware) > self->major_vcn) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "image VCN is not compatible, got 0x%x, expected 0x%x",
+			    fu_igsc_aux_firmware_get_major_vcn(firmware),
+			    self->major_vcn);
+		return NULL;
+	}
+	if (fu_igsc_aux_firmware_get_oem_version(firmware) <= self->oem_version) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "invalid manufacturer data version, got 0x%x, expected 0x%x",
+			    fu_igsc_aux_firmware_get_oem_version(firmware),
+			    self->oem_version);
+		return NULL;
+	}
+
+	/* success, but return container, not CPD */
+	return FU_FIRMWARE(g_steal_pointer(&firmware));
+}
+
+static gboolean
+fu_igsc_aux_device_write_firmware(FuDevice *device,
+				  FuFirmware *firmware,
+				  FuProgress *progress,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	FuIgscDevice *igsc_parent = FU_IGSC_DEVICE(fu_device_get_parent(device));
+	g_autoptr(GBytes) fw_info = NULL;
+	g_autoptr(GBytes) fw_payload = NULL;
+
+	/* get image */
+	fw_info =
+	    fu_firmware_get_image_by_idx_bytes(firmware, FU_IFWI_FPT_FIRMWARE_IDX_INFO, error);
+	if (fw_info == NULL)
+		return FALSE;
+	fw_payload =
+	    fu_firmware_get_image_by_idx_bytes(firmware, FU_IFWI_FPT_FIRMWARE_IDX_SDTA, error);
+	if (fw_payload == NULL)
+		return FALSE;
+	return fu_igsc_device_write_blob(igsc_parent,
+					 GSC_FWU_HECI_PAYLOAD_TYPE_FWDATA,
+					 fw_info,
+					 fw_payload,
+					 progress,
+					 error);
+}
+
+static gboolean
+fu_igsc_aux_device_prepare(FuDevice *device,
+			   FuProgress *progress,
+			   FwupdInstallFlags flags,
+			   GError **error)
+{
+	/* set PCI power policy */
+	return fu_device_prepare(fu_device_get_parent(device), progress, flags, error);
+}
+
+static gboolean
+fu_igsc_aux_device_cleanup(FuDevice *device,
+			   FuProgress *progress,
+			   FwupdInstallFlags flags,
+			   GError **error)
+{
+	/* set PCI power policy */
+	return fu_device_cleanup(fu_device_get_parent(device), progress, flags, error);
+}
+
+static void
+fu_igsc_aux_device_init(FuIgscAuxDevice *self)
+{
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
```

Comment:
the card doesn't need to restart or anything?

Reply:
That's the bit that doesn't work yet. It might, and I don't know what actually happens to the displays connected to the updated GPU. It might be NEEDS_REBOOT.

Reply:
Yeah depending how this all works under the covers I think you need to characterize the update behavior then we'll before this is merged.

Reply:
I've removed _USABLE_DURING_UPDATE -- waiting for firmware files before checking NEEDS_REBOOT.